### PR TITLE
feat: support live development and deployment for Remote App projects

### PR DESCRIPTION
**Pull request by izaera**
See https://issues.liferay.com/browse/LPS-160803

### CODE REVIEW

**Comment by izaera** [comment]
I've fixed the QA scripts, but I'm not sure if the `yarn qa check` will work because it tries to compare the output ZIP files by hashing them but, as file times will change, it will probably fail in future runs. 

So I'll add a commit to check the internals of the ZIP file file by file.

**Comment by izaera** [comment]
I've added support for diffing the internals of ZIP files in the QA scripts. This is ready for merge now.

**Comment by izaera** [comment]
Converted to draft again because I want to add a `--only` switch to the `start` command so that the live development server can be launched without automatic deployment to liferay being executed.

File: projects/js-toolkit/packages/portal-base/src/build/bundler2.ts [hunk 1]
```diff
@@ -106,7 +106,14 @@ function runBabel(project: Project): void {
 }
 
 function runCompiler(project: Project): void {
-	runBabel(project);
+	const dependencies = project.pkgJson.dependencies || {};
+
+	if (dependencies['@angular/core']) {
```
Comment (bryceosterhaus) [inline projects/js-toolkit/packages/portal-base/src/build/bundler2.ts:111]: 👍

File: projects/js-toolkit/packages/portal-base/src/build/customElement.ts [hunk 2]
```diff
@@ -25,6 +34,42 @@ export default async function customElement(project: Project): Promise<void> {
...
 	await runWebpack(project, configuration);
+
+	await makeZip(project);
+}
+
+/**
+ * Add several files to a ZIP folder.
+ * @param srcDirPath source folder
+ * @param srcGlobs array of globs describing files to include (in
+ * 						globby, i.e. POSIX, format)
```
Comment (bryceosterhaus) [inline projects/js-toolkit/packages/portal-base/src/build/customElement.ts:45]: whitespace looks a little odd here

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts [hunk 3]
```diff
@@ -37,4 +54,108 @@ export default async function customElement(project: Project): Promise<void> {
...
+}
+
+async function checkForOldZipFile(zipFile: FilePath): Promise<void> {
+	if (!fs.existsSync(zipFile.asNative)) {
+		return;
+	}
+
+	const stats = fs.statSync(zipFile.asNative);
+
+	if (Date.now() - stats.mtimeMs > 1000 * 60 * 120) {
```
Comment (bryceosterhaus) [inline projects/js-toolkit/packages/portal-base/src/start/customElement.ts:76]: Can we set `1000 * 60 * 120` to a variable with a explicit name?

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts [hunk 4]
```diff
@@ -37,4 +54,108 @@ export default async function customElement(project: Project): Promise<void> {
...
+		const lines = createInterface({
+			input: process.stdin,
+		});
+
+		print(warn`
+There is an older than 2 hours version of this project in your Liferay server.`);
```
Comment (bryceosterhaus) [inline projects/js-toolkit/packages/portal-base/src/start/customElement.ts:82]: Why the 2 hour limit? Was this just arbitrary?

File: projects/js-toolkit/packages/portal-base/src/util/promptForDeployPath.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+): Promise<FilePath> {
+	const lines = createInterface({
+		input: process.stdin,
+	});
+
+	print(question`Please enter your local Liferay installation directory`);
+
+	let deployDir: FilePath;
+
+	for await (const line of lines) {
```
Comment (bryceosterhaus) [inline projects/js-toolkit/packages/portal-base/src/util/promptForDeployPath.ts:23]: Whoa, haven't seen `for await` before, very cool

**Review by bryceosterhaus — APPROVED** [review]
LGTM! I had a few small comments but looks fine by me.

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts (hunk 4 above)
Comment (izaera) [reply projects/js-toolkit/packages/portal-base/src/start/customElement.ts:82]: Yes. It seems reasonable to me. My intention is to warn only in cases where you may have deployed the application some time ago, in a previous development session, but not every time you run `yarn start`. Two hours seems enough to not trigger false positives when you stop for lunch :sweat_smile: 

However, I'm now thinking that people won't be running `yarn start` a lot of times. Chances are that they just run it once per development session, so maybe we can warn always (without checking for the elapsed time)... :thinking:

File: projects/js-toolkit/packages/portal-base/src/util/promptForDeployPath.ts (hunk 5 above)
Comment (izaera) [reply projects/js-toolkit/packages/portal-base/src/util/promptForDeployPath.ts:23]: you don't know about it until you need it :sweat_smile:

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts (hunk 4 above)
Comment (bryceosterhaus) [reply projects/js-toolkit/packages/portal-base/src/start/customElement.ts:82]: Sounds good, just wanted to make sure I fully understood!

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts (hunk 3 above)
Comment (izaera) [reply projects/js-toolkit/packages/portal-base/src/start/customElement.ts:76]: for example, something like:

```javascript
const ONE_WEEK = 1000 * 60 * 120;
```

:trollface:

File: projects/js-toolkit/packages/portal-base/src/start/customElement.ts (hunk 4 above)
Comment (izaera) [reply projects/js-toolkit/packages/portal-base/src/start/customElement.ts:82]: I'm going to remove the check for the two hours...

**Pull request merged**